### ISSUE TITLE
Better support for multiple vhost files

### ISSUE THREAD

**Issue by strategist333** [geerlingguy/ansible-role-nginx#70]
Per https://github.com/ansible/ansible/issues/12623, `with_items` is not supported with roles in Ansible2. However, the current setup of the code is strongly geared toward a single `nginx_vhosts_filename`, and indeed the readme says

```
If you run the role multiple times (e.g. include the role with with_items), you can change the name for each run, effectively creating a separate vhosts file per vhost configuration.
```

The current recommendation seems to be to move the `with_items` into the role.

**Comment by geerlingguy** [comment]
@strategist333 - Can you do:

```yaml
  roles:
    - role: geerlingguy.nginx
      param1: test
    - role: geerlingguy.nginx
      param2: test
```

I'm guessing not, based on the way the object would be structured... but it might work. If so, that would be the best workaround for now.

**Comment by strategist333** [comment]
That actually works, thanks! However, there is the caveat that:

1. one must hard-code the number of vhosts into the roles block
2. there is a lot of duplicate work that is done as all of the non-vhost-related tasks are re-executed
3. if any vhosts are removed, the corresponding file isn't cleaned up. (When I first went to try this, the nginx reload was failing because the vhosts.conf file was left over from before, leading to conflicting server blocks.) This is a separate issue, but running the role many times prevents the simple resolution of clearing the entire `nginx_vhost_path` folder.

**Comment by geerlingguy** [comment]
@strategist333 - thus the "workaround for now" :D

I'm thinking it's worth making the server blocks/virtual hosts configuration a bit nicer... basically, allow an array of server definitions, and then save one file per definition, deleting any that aren't in the list... something like that.

However, the auto-deletion of nonexistent ones could be tricky, and definitely should be disabled by default (just to make sure it doesn't mess up existing setups that may have extra conf files).

**Comment by jszaszvari** [comment]
This would be awesome, I've been trying to figure out how I can deploy a web server with multiple nginx vhost's now for a few weeks and still cant get it :(

**Comment by babatundebusari** [comment]
@geerlingguy 
any new developments on this feature?
for multiple file-based vhost configuration

**Comment by mekonhead** [comment]
You can create individual vhost config files with the following:
```
- name: Add managed vhost config file (if any vhosts are configured).
  template:
    src: "{{ item.src }}"
    dest: "{{ nginx_vhost_path }}/{{ item.dest }}"
    mode: 0644
  with_items: "{{ nginx_vhost_list }}"
  when: nginx_vhosts|length > 0
  notify: reload nginx
```
And then use the following list:
```
nginx_vhost_list:
  - src: vhosts.j2
    dest: server01.conf
  - src: vhosts.j2
    dest: server02.conf
```
The issue then is to add the server name into the template so each vhost config has the correct servername directive.

You could add another item to the list such as follows:
```
nginx_vhost_list:
  - src: vhosts.j2
    dest: server01.conf
    servername: server01
  - src: vhosts.j2
    dest: server02.conf
    servername: server02
```
But how do you then reference that in the template so each iteration gets the correct servername ?

If you do the following you just get both names in each vhost file:
```
{% for vhost in nginx_vhost_list %}
{% if vhost.servername is defined %}
    server_name {{ vhost.servername }};
{% endif %}
{% endfor %}
```
I'm a bit new to this so any help would be appreciated. Being able to create multiple vhost files would be very handy.

**Comment by mekonhead** [comment]
You can add the config files from a dictionary:
```
nginx_vhost_dict:
  vhost01:
    servername: vhost01
    domain: company.com
  vhost02:
    servername: vhost02
    domain: company.com
```
Using something similar to the following:
```
- name: Add managed vhost config files.
  template:
    src: vhost.j2
    dest: "{{ nginx_vhost_path }}/{{ item.key }}.conf"
    mode: 0644
  with_dict: "{{ nginx_vhost_dict }}"
  notify: reload nginx
```
However when it comes to removing unmanaged config files that are no longer in the dict the following does not error but does not actually remove them:
```
- name: Remove unmanaged vhost config files (if vhost is not configured).
  file:
    path: "{{ nginx_vhost_path }}/{{ item.key }}.conf"
    state: absent
  with_dict: "{{ nginx_vhost_dict }}"
  when: item.key not in nginx_vhost_dict
  notify: reload nginx
```
Any thoughts as to why not and how to do this would be greatly appreciated.

**Comment by fxer** [comment]
I might also add sometimes you want 2 vhosts per file, such as for SSL when the "listen 80" server block is simply a 301 redirect to your "listen 443" server block, which contains most of the config

**Comment by geerlingguy** [comment]
I would like to get this fixed up (see some more rationale in #1255)—I currently have a setup on my _own_ servers like:

```yaml
- name: Copy Nginx vhosts into place.
  template:
    src: vhost.conf.j2
    dest: "/etc/nginx/sites-enabled/{{ item.server_name_canonical }}.conf"
    force: yes
    owner: root
    group: root
    mode: 0644
  with_items: "{{ nginx_vhosts }}"
  notify: restart nginx
```

And it uses a variable structure like:

```yaml
mm_nginx_vhosts:
  - server_name_canonical: www.example.com
    server_name_redirect: example.com
    # Note: Setting this value triggers some extra template logic for Drupal.
    drupal_docroot: /var/www/domains/d7multisite
    ssl_enabled: true
    ssl_certificate: /etc/letsencrypt/live/www.example.com/fullchain.pem
    ssl_key: /etc/letsencrypt/live/www.example.com/privkey.pem
    extra_parameters: |
      # Extra configuration for Midwestern Mac.
          location = /path/to/redirect {
              return 301 https://example.com/redirected/path;
          }
```

Then the template has all the magic inside. If the template were extensible, it would be easier for downstream projects like Drupal VM to add in configuration like that which is required for `drupal_docroot`. I also use that flag to pass PHP-FPM through a server on port 81 that's then cached on port 80.

**Comment by geerlingguy** [comment]
Fixed via #91

**Comment by jszaszvari** [comment]
Thanks @geerlingguy 

Great stuff
